### DISCUSSION TITLE
Plotting 3D vector field with quiver3

### DISCUSSION THREAD

**Discussion by matthew-hennefarth** [alandefreitas/matplotplusplus#144] [Q&A] [answered]
I am looking to plot 3D vector fields, though I am having trouble doing so. When I try a very simple example such as:

    #include <vector>
    #include <matplot/matplot.h>
    
    int main() {
    
      //specifies location
      std::vector<double> x;
      std::vector<double> y;
      std::vector<double> z;
    
      //represent the field
      std::vector<double> ex;
      std::vector<double> ey;
      std::vector<double> ez;
    
      x.push_back(1.0);
      y.push_back(1.0);
      z.push_back(0.0);
    
      ex.push_back(1.0);
      ey.push_back(0.0);
      ez.push_back(0.0);
    
      matplot::quiver3(x, y, z, ex, ey, ez);
      matplot::show();
      return 0;
    }

the gnuplot window pops up, but nothing is shown and I get the following error: 

    multiplot> splot  '-' with vectors  linecolor rgb "0x000071BC" dashtype 1 linewidth 0.5 linetype -1
                                                                                                       ^
               line 2: All points x value undefined
    
    
    multiplot> splot  '-' with vectors  linecolor rgb "0x000071BC" dashtype 1 linewidth 0.5 linetype -1
                                                                                                       ^
               line 4: All points x value undefined

I am not sure if I am calling the function correctly, or what I am doing wrong; but I would love to get this working! Let me know what I need to fix to get this working!

**Comment by alandefreitas** [comment, marked as answer]
Does [this example](https://github.com/alandefreitas/matplotplusplus/blob/master/examples/vector_fields/quiver3/quiver3_1.cpp) work? It might serve as a starting point.

**Reply by matthew-hennefarth** [reply]
I had read through it, but the function call is using the Z surface, with an implicitly defined x,y coordinate scheme. I have explicit x,y,z positions for each vector. Though, I was able to get it working by normalizing my vectors first, and then scaling them!

Also, I noticed that there is no colormap feature for vectors at the moment (please correct me if I am wrong) and I have forked the repo and am working on adding this right now. I can submit a pull request to add this feature, let me know!

**Reply by alandefreitas** [reply]
> I had read through it, but the function call is using the Z surface, with an implicitly defined x,y coordinate scheme. I have explicit x,y,z positions for each vector. Though, I was able to get it working by normalizing my vectors first, and then scaling them!

I see. Good to know it worked somehow. It's been a while since these plot types were implemented so I'd have to check the code and come up with some examples to be sure about these problems.

> Also, I noticed that there is no colormap feature for vectors at the moment (please correct me if I am wrong) and I have forked the repo and am working on adding this right now. I can submit a pull request to add this feature, let me know!

That's true. Sure. PRs are more than welcome.
